core/src/main/java/org/infinispan/remoting/inboundhandler/TxNonTotalOrderPerCacheInboundInvocationHandler.java

```diff
@@ -0,0 +1,114 @@
+
+   private static final Log log = LogFactory.getLog(TxNonTotalOrderPerCacheInboundInvocationHandler.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   @Override
+   public void handle(CacheRpcCommand command, Reply reply, DeliverOrder order) {
+      if (order == DeliverOrder.TOTAL) {
+         unexpectedDeliverMode(command, order);
+      }
+      try {
+         final boolean onExecutorService;
+         final BlockingRunnable runnable;
+         final int commandTopologyId;
+
+         switch (command.getCommandId()) {
+            case SingleRpcCommand.COMMAND_ID:
+               commandTopologyId = extractCommandTopologyId((SingleRpcCommand) command);
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
+                  reply.reply(CacheNotFoundResponse.INSTANCE);
+                  return;
+               }
+               onExecutorService = executeOnExecutorService(order, command);
+               runnable = createDefaultRunnable(command, reply, commandTopologyId, true, onExecutorService);
+               break;
+            case MultipleRpcCommand.COMMAND_ID:
+               commandTopologyId = extractCommandTopologyId((MultipleRpcCommand) command);
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
+                  reply.reply(CacheNotFoundResponse.INSTANCE);
+                  return;
+               }
+               onExecutorService = executeOnExecutorService(order, command);
+               runnable = createDefaultRunnable(command, reply, commandTopologyId, true, onExecutorService);
+               break;
+            case StateRequestCommand.COMMAND_ID:
+               // StateRequestCommand is special in that it doesn't need transaction data
+               // In fact, waiting for transaction data could cause a deadlock
+               commandTopologyId = extractCommandTopologyId((StateRequestCommand) command);
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
+                  reply.reply(CacheNotFoundResponse.INSTANCE);
+                  return;
+               }
+               onExecutorService = executeOnExecutorService(order, command);
+               runnable = createDefaultRunnable(command, reply, commandTopologyId, false, onExecutorService);
+               break;
+            case PrepareCommand.COMMAND_ID:
+            case VersionedPrepareCommand.COMMAND_ID:
+               commandTopologyId = extractCommandTopologyId((PrepareCommand) command);
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
+                  reply.reply(CacheNotFoundResponse.INSTANCE);
+                  return;
+               }
+               onExecutorService = executeOnExecutorService(order, command);
+               runnable = createDefaultRunnable(command, reply, commandTopologyId, true, onExecutorService);
+               break;
+            case LockControlCommand.COMMAND_ID:
```

Comment:
the lock and the prepare are the same as default now, but it will change as soon as I integrate the rest of the modifications.

I'll try to find a way to remove the duplicated code.